sql/src/main/java/io/crate/planner/operators/FetchOrEval.java

```diff
@@ -103,34 +105,59 @@
     final List<Symbol> outputs;
 
     private final FetchMode fetchMode;
-    private final boolean isLastFetch;
+    private final boolean doFetch;
 
     static LogicalPlan.Builder create(LogicalPlan.Builder sourceBuilder,
                                       List<Symbol> outputs,
                                       FetchMode fetchMode,
-                                      boolean isLastFetch) {
+                                      boolean isLastFetch,
+                                      boolean childIsLimited) {
         return (tableStats, usedBeforeNextFetch) -> {
             final LogicalPlan source;
-            if (fetchMode == FetchMode.NEVER || !isLastFetch) {
+
+            boolean doFetch = isLastFetch;
+            if (fetchMode == FetchMode.NEVER_CLEAR) {
                 source = sourceBuilder.build(tableStats, usedBeforeNextFetch);
-            } else {
+            } else if (isLastFetch) {
                 source = sourceBuilder.build(tableStats, Collections.emptySet());
+            } else {
+                /*
+                 * In a case like
+                 *      select sum(x) from (select x from t limit 10)
+                 *
+                 * It makes sense to do an intermediate fetch, to reduce the amount of rows.
+                 * All columns are used, so a _fetchId propagation wouldn't work.
+                 *
+                 *
+                 * But in a case like
+                 *      select x, y from (select x, y from t order by x limit 10) order by x asc limit 5
+                 *
+                 * A _fetchId propagation makes sense because there are unusedColumns (y), which can be fetched
+                 * at the end.
+                 */
+                List<Symbol> unusedColumns = getUnusedColumns(outputs, usedBeforeNextFetch);
```

Comment:
can we extract the `getUnusedColumns` into a shared class? looks a bit odd to me to use static methods from the collect operation. otherwise pls remove at least the visibleForTesting annotation.

Reply:
Pushed a second commit